client-rpc/src/rpc/websocket_core.rs

```diff
@@ -0,0 +1,382 @@
+use serde_json::json;
+use serde_json::Value;
+use std::sync::mpsc;
+use std::time;
+use std::time::SystemTime;
+use websocket::OwnedMessage;
+
+// finite state machine that manages blocks
+// just use one thread to multi-plexing for data and command
+
+// rust don't allow sharing between threads without mutex
+// so multi-plexed with OwnedMessage
+
+// Network is handled websocket_rpc 
+
+// not to use too much cpu, it takes some time for waiting
+const WAIT_PROCESS_TIME: u128 = 5000; // milli seconds
+const BLOCK_REQUEST_TIME: u128 = 10; // milli seconds
+const RECEIVE_TIMEOUT: u64 = 10; //  milli seconds
+#[derive(Copy, Clone, Debug)]
+pub enum WebsocketState {
+    ReadyProcess,
+    GetStatus,
+    GetBlocks,
+    WaitProcess,
+}
+
+pub struct WebsocketCore<S, C, H, T>
+where
+    S: Storage,
+    C: Client,
+    H: BlockHandler,
+    T: WalletClient,
+{
+    sender: futures::sync::mpsc::Sender<OwnedMessage>,
+    my_sender: Sender<OwnedMessage>,
+    my_receiver: Receiver<OwnedMessage>,
+    old_blocktime: SystemTime,
+    state_time: SystemTime,
+
+    max_height: u64,
+    state: WebsocketState,
+    global_state_service: GlobalStateService<S>,
+    client: C,
+    block_handler: H,
+    wallets: Vec<WalletInfo>,
+    current_wallet: usize,
+    wallet_client: T,
+}
+
+impl<S, C, H, T> WebsocketCore<S, C, H, T>
+where
+    S: Storage,
+    C: Client,
+    H: BlockHandler,
+    T: WalletClient,
+{
+    pub fn new(
+        sender: futures::sync::mpsc::Sender<OwnedMessage>,
+        storage: S,
+        client: C,
+        block_handler: H,
+        wallets: WalletInfos,
+        wallet_client: T,
+    ) -> Self {
+        let gss = GlobalStateService::new(storage);
+
+        let channel = mpsc::channel();
+        // tx, rx
+        let (my_sender, my_receiver) = channel;
+
+        WebsocketCore {
+            sender,
+            my_sender,
+            my_receiver,
+            old_blocktime: SystemTime::now(),
+            state_time: SystemTime::now(),
+
+            max_height: 0,
+            state: WebsocketState::ReadyProcess,
+            global_state_service: gss,
+            client,
+            block_handler,
+            wallets,
+            current_wallet: 0,
+
+            wallet_client,
+        }
+    }
+
+    // to process multiple wallets
+    fn get_current_wallet(&self) -> WalletInfo {
+        assert!(self.current_wallet < self.wallets.len());
+        self.wallets[self.current_wallet].clone()
+    }
+
+    // get height from database 
+    fn get_current_height(&self) -> u64 {
+        let wallet = self.get_current_wallet();
+        self.global_state_service
+            .last_block_height(&wallet.view_key)
+            .unwrap()
+    }
+
+    // write block to internal database
+    fn do_save_block_to_chain(&mut self, value: &Value, kind: &str) {
+        if self.wallets.is_empty() {
+            return;
+        }
+        // restore as object
+        let height: u64 = value["block"]["header"]["height"]
```

Comment:
Use `Block` struct in `client-common` instead of parsing manually.

Reply:
ok